src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,433 @@
+         * creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt
+         * but the actual value of the salt is not yet available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(byte[] salt) {
+            if(salt == null) {
+                throw new NullPointerException(
+                    "salt must not be null or empty");
+            }
+            if (salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Returns a builder for building {@code Extract} and {@code ExtractExpand} objects.
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should
+     * be called next, before calling build methods, such as
+     * {@code Builder.extractOnly()}
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder builder() {
+        return new Builder().createExtract();
+    }
+
+    /**
+     * Static helper-method that may be used to initialize an {@code Expand}
+     * object
+     *
+     * @param prk
+     *     the pseudorandom key
+     * @param info
+     *     the optional context and application specific information
+     * @param length
+     *     the length of the output key material (must be > 0)
+     *
+     * @return a new {@code Expand} object
+     *
+     * @throws NullPointerException
+     *     if {@code prk} or {@code info} is {@code null}
+     */
+    static Expand expand(SecretKey prk, byte[] info, int length) {
+        if (prk == null) {
+            throw new NullPointerException("prk must not be null");
+        }
+        if (info == null) {
+            throw new NullPointerException("info must not be null");
+        }
+        return new Expand(prk, info, length);
+    }
+
+    /**
+     * Defines the input parameters of an Extract operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Extract implements HKDFParameterSpec {
+
+        // HKDF-Extract(salt, IKM) -> PRK
+        private final List<SecretKey> ikms;
+        private final List<SecretKey> salts;
+
+        private Extract() {
+            this(new ArrayList<>(), new ArrayList<>());
+        }
+
+        private Extract(List<SecretKey> ikms, List<SecretKey> salts) {
+            this.ikms = List.copyOf(ikms);
+            this.salts = List.copyOf(salts);
+        }
+
+        /**
+         * Returns an unmodifiable {@code List} of initial key material
+         * values.
```

Comment:
Add a sentence like "an implementation should concatenate the input key materials into a single byte array and supply it into the HKDF-Extract as defined in section n.n of RFC xxxx". Same for `salts`.